### DISCUSSION TITLE
how to expand variable as command parameters?

### DISCUSSION THREAD

**Discussion by JeffMelton** [lmorg/murex#901] [Q&A] [answered]
New Murex user here, just started tinkering with it tonight. I'm an inveterate user of oddball shells (currently using Elvish as my daily). I'm trying and failing to replicate some of my Elvish setup, really just failing to understand some basics. I use `eza`, and I borrowed an idea from a Fish plugin where I set environment variables to some of the common `eza` flags I'd use, then use those in some functions. My current Murex function looks like this:

```murex
function eza_git {
    if { git <!null> rev-parse --is-inside-work-tree } then {
        command eza @ENV.EZA_STANDARD_OPTIONS @ENV.EZA_LL_OPTIONS --git $@
    } else {
        command eza @ENV.EZA_STANDARD_OPTIONS @ENV.EZA_LL_OPTIONS $@
    }
}
```
 I've got this set in `.murex_preload`:
 
 ```murex
 export EZA_STANDARD_OPTIONS = "--group --header --group-directories-first"
export EZA_LA_OPTIONS = "--all --binary"
export EZA_LL_OPTIONS = "--long"
```

So far, everything I've tried returns something along these lines:

```
eza: Unknown argument --group --header --group-directories-first
```

So what I'd like is for `eza` to be able to use those variables as flags, but so far I'm not getting it done. If someone can point me in the right direction I'd appreciate it! :)

**Comment by lmorg** [comment, marked as answer]
Welcome 👋 

Murex doesn't auto-split strings by whitespace by default.

There's a couple of ways you can solve this. If `EZA_STANDARD_OPTIONS` (et al) doesn't need to be environmental variables (ie passed as values to child processes outside of Murex) then you can store them as global variables and thus they can be typed as an array. eg

```
EZA_STANDARD_OPTIONS = %[ --group --header --group-directories-first ]
EZA_LA_OPTIONS       = %[ --all --binary ]
EZA_LL_OPTIONS       = %[ --long ]
```
(https://murex.rocks/parser/create-array.html#strings)

Then your function could look like:

```
function eza_git {
    if { git <!null> rev-parse --is-inside-work-tree } then {
        command eza @EZA_STANDARD_OPTIONS @EZA_LL_OPTIONS --git @PARAMS
    } else {
        command eza @EZA_STANDARD_OPTIONS @EZA_LL_OPTIONS @PARAMS
    }
}
```

(I'm guessing one of the `EZA_LL_OPTIONS` above should be `LA`?)

Also `$@` doesn't exist. You'll probably want `@PARAMS`:
https://murex.rocks/variables/params.html

You might also need to disable empty array checks too. They exist so that commands fail if an array is empty rather than running potentially unsafe commands due to expected parameters being missing. This can be disabled via
```
config set proc strict-arrays false
```

For example there is a `bash` wrapper function that disables syntax highlighting in Murex (it causes `readline` issues when Bash is invoked inside Murex) that looks like:
```
function bash {
    # Wrapper script around `bash`

    config set proc strict-arrays false
    config set proc force-tty true
    exec bash @PARAMS
}
```
So that parameters are passed to Bash but also Bash can still execute even when `@PARAMS` is empty. However you might prefer `eza_git` to fail if any of your arrays are empty?

It's also worth noting that adding a comment to the top of your function then adds a description to Murex. For example the hint text at the bottom of this screenshot was provided by the comment inside the function:
![image](https://github.com/user-attachments/assets/16504840-011a-46fd-87ec-66822567556b)
I don't appear to have that feature documented though, so I'll get that added to the docs

**Reply by lmorg** [reply]
If you'd prefer to have your `EZA` variables set as environmental variables then you can convert them to arrays via `jsplit`, eg
```
@{$ENV.VAR|jsplit \s}
```
https://murex.rocks/commands/jsplit.html

But I'll be the first to admit that this is arguably pretty ugly code. Though you could technically store the string version as a env var and the array version as a global (they wouldn't overwrite each other) so each context has a value that best supports each platform....if you specifically do need to pass values to child processes that is.

**Reply by JeffMelton** [reply]
Super helpful, thank you. There's a *ton* of documentation, and that's good, but it's a lot to ingest. 😀

**Reply by lmorg** [reply]
On the topic of documentation, I'm always looking for ways to improve it. So if you have any thoughts there then I'm open to suggestions 👍

**Reply by JeffMelton** [reply]
Oh, I haven't spent nearly enough time in it yet to have informed opinions. It's unfamiliar, but that's expected. I'll keep it in mind, though, thanks.
